### DISCUSSION TITLE
using django template engine?

### DISCUSSION THREAD

**Discussion by araneta** [a-h/templ#616] [General]
is it possible to create a templates that support extends and block like django template engine?
{% extends "layouts/admin.html" %}
{% block innercontent %}
{% endblock %}
or maybe we can user https://github.com/flosch/pongo2 ?

**Comment by joerdav** [comment]
This is supported yes, but the syntax is very different. And it's a bit of a paradigm shift. I think what might be best is if you show a template in django style, I'll show you how I would architect that in templ.

**Comment by araneta** [comment]
here is the sample of layouts/main.html

```
<!doctype html>
<head>
{%block css %}{% endblock %}
</head>
<body class="bg-gradient-primary">
 {% block content %} {% endblock %}
</body>
</html>

```
and the page can inherit that layout for example login.html

```
{% extends "layouts/main.html" %}
{% block content %}
<div class="container">
test content
</div>
{% endblock %}

```
how do i implement this in templ ? thanks

**Reply by joerdav** [reply]
Ah, so with templ you may have to invert your thinking a little bit, but this is how:

```
package layouts

templ layout() {
    <!doctype html>
    <html>
        <body class="bg-gradient-primary">
                 { children... }
        </body>
    </html>
}

templ mainPage() {
    @layout() {
        <div class="container">
            test content
        </div>
    }
}
```

**Comment by araneta** [comment]
i see but there are multiple blocks in the layouts/main.html like css and content how can i do that?
thanks

**Reply by joerdav** [reply]
In the case of HTML you can use function arguments to render components. 

Please read the page about composition: https://templ.guide/syntax-and-usage/template-composition

```
package layouts

templ layout(contents, css templ.Component) {
    <!doctype html>
    <html>
        <head>
                @css
        </head>
        <body class="bg-gradient-primary">
                 @contents
        </body>
    </html>
}

templ mainPage() {
    @layout(mainContents(), mainCSS())
}

templ mainContents() {
        <div class="container">
            test content
        </div>
}

templ mainCSS() {
        <styles>
            ...
        </styles>
}
```

**Comment by araneta** [comment]
thanks
